soco/music_services/music_service.py

```diff
@@ -155,50 +196,138 @@ def call(self, method, args=None):
         try:
             result_elt = message.call()
         except SoapFault as exc:
+            if "Client.AuthTokenExpired" in exc.faultcode:
+                raise MusicServiceAuthException(
+                    "Authorization for {} expired or invalid: [{} / {} / {}]".format(
+                        self.music_service.service_name,
+                        exc.faultcode,
+                        exc.faultstring,
+                        exc.detail,
+                    )
+                ) from exc
+
             if "Client.TokenRefreshRequired" in exc.faultcode:
-                log.debug("Token refresh required. Trying again")
+                log.debug(
+                    "Auth token for %s expired, attempting to refresh",
+                    self.music_service.service_name,
+                )
+                if self.music_service.auth_type not in ("DeviceLink", "AppLink"):
+                    raise MusicServiceAuthException(
+                        "Token-refresh not supported for music service auth type: "
+                        + self.music_service.auth_type
+                    ) from exc
+
                 # Remove any cached value for the SOAP header
                 self._cached_soap_header = None
 
-                # <detail>
-                #   <refreshAuthTokenResult>
-                #       <authToken>xxxxxxx</authToken>
-                #       <privateKey>zzzzzz</privateKey>
-                #   </refreshAuthTokenResult>
+                # Extract new token and key from the error message
+                # <detail xmlns:ms="http://www.sonos.com/Services/1.1">
+                #   <ms:RefreshAuthTokenResult>
+                #     <ms:authToken>xxxxxxx</ms:authToken>
+                #     <ms:privateKey>yyyyyy</ms:privateKey>
+                #   </ms:RefreshAuthTokenResult>
                 # </detail>
                 auth_token = exc.detail.findtext(".//authToken")
                 private_key = exc.detail.findtext(".//privateKey")
-                # We have new details - update the account
-                self.music_service.account.oa_device_id = auth_token
-                self.music_service.account.key = private_key
+
+                # Create new token pair and save it
+                token_pair = (auth_token, private_key)
+                self.token_store.save_token_pair(
+                    self.music_service.service_id,
+                    self._device.household_id,
+                    token_pair,
+                )
```

Comment:
If the payload does not contain a token/key pair could the token store save empty values?

Reply:
Good point. Fixed in 976922d.